### PR TITLE
8240349: jlink should not leave partial image output directory on failure

### PR DESCRIPTION
**Pull request by sundararajana**
jlink should clean up output directory on any failure. should not leave partially filled output.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8240349](https://bugs.openjdk.java.net/browse/JDK-8240349): jlink should not leave partial image output directory on failure

### Reviewers
 * [Jim Laskey](https://openjdk.java.net/census#jlaskey) (@JimLaskey - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4386/head:pull/4386` \
`$ git checkout pull/4386`

Update a local copy of the PR: \
`$ git checkout pull/4386` \
`$ git pull https://git.openjdk.java.net/jdk pull/4386/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4386`

View PR using the GUI difftool: \
`$ git pr show -t 4386`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4386.diff">https://git.openjdk.java.net/jdk/pull/4386.diff</a>

</details>

### CODE REVIEW

**Review by JimLaskey — APPROVED** [review]

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java [hunk 1]
```diff
@@ -285,6 +290,7 @@ int run(String[] args) {
             if (DEBUG) {
                 e.printStackTrace(log);
             }
+            cleanupOutput(outputPath);
             return EXIT_ERROR;
         } catch (BadArgs e) {
```
Comment (AlanBateman) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:295]: Can you confirm that this change is benign for the case that the output directory exists? I believe it's a BadArgs case so it won't attempt to delete a file or tree if it exists but I'd like confirmation.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java [hunk 2]
```diff
@@ -468,6 +486,30 @@ public static ModuleFinder newModuleFinder(List<Path> paths,
         return finder;
     }
 
+    private static void deleteDirectory(Path dir) throws IOException {
+        if (dir != null && Files.isDirectory(dir)) {
```
Comment (AlanBateman) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:490]: I don't think this method should be checking dir, that should be up to the caller to ensure that it is not called when output is null.

**Review by AlanBateman — COMMENTED** [review]
I think we need to rename the JBS issue as this is not specific to the -vm option. Instead, this is about cleanup when jlink fails with some intermediate files in the output directory.

**Comment by JornVernee** [comment]
WRT the test failure on Windows discussed offline: when the directory is deleted as a result of a `PluginException` being thrown, there is still an open file handle on the `lib/modules` file in the image directory, which prevents the directory from being deleted.

Bisecting this, it seems that the file handle is being created in `ImageFileCreater::writeImage` with a call to `plugins.getJImageFileOutputStream` (https://github.com/openjdk/jdk/blob/master/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java#L162). This creates an output stream that is never closed.

Following patch fixes:
```
diff --git a/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java b/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java
index 749025bea9d..8beddc5a037 100644
--- a/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java
+++ b/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java
@@ -158,8 +158,10 @@ public final class ImageFileCreator {
         BasicImageWriter writer = new BasicImageWriter(byteOrder);
         ResourcePoolManager allContent = createPoolManager(archives,
                 entriesForModule, byteOrder, writer);
-        ResourcePool result = generateJImage(allContent,
-             writer, plugins, plugins.getJImageFileOutputStream());
+        ResourcePool result;
+        try (DataOutputStream out = plugins.getJImageFileOutputStream()) {
+            result = generateJImage(allContent, writer, plugins, out);
+        }

         //Handle files.
         try {
```

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java (hunk 1 above)
Comment (sundararajana) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:295]: Will check that.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java (hunk 2 above)
Comment (sundararajana) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:490]: okay. will move that check to caller.

**Comment by sundararajana** [comment]
> WRT the test failure on Windows discussed offline: when the directory is deleted as a result of a `PluginException` being thrown, there is still an open file handle on the `lib/modules` file in the image directory, which prevents the directory from being deleted.
> 
> Bisecting this, it seems that the file handle is being created in `ImageFileCreater::writeImage` with a call to `plugins.getJImageFileOutputStream` (https://github.com/openjdk/jdk/blob/master/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java#L162). This creates an output stream that is never closed.
> 
> Following patch fixes:
> 
> ```
> diff --git a/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java b/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java
> index 749025bea9d..8beddc5a037 100644
> --- a/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java
> +++ b/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java
> @@ -158,8 +158,10 @@ public final class ImageFileCreator {
>          BasicImageWriter writer = new BasicImageWriter(byteOrder);
>          ResourcePoolManager allContent = createPoolManager(archives,
>                  entriesForModule, byteOrder, writer);
> -        ResourcePool result = generateJImage(allContent,
> -             writer, plugins, plugins.getJImageFileOutputStream());
> +        ResourcePool result;
> +        try (DataOutputStream out = plugins.getJImageFileOutputStream()) {
> +            result = generateJImage(allContent, writer, plugins, out);
> +        }
> 
>          //Handle files.
>          try {
> ```

Thanks @JornVernee

**Review by JimLaskey — APPROVED** [review]

**Review by AlanBateman — APPROVED** [review]
New version looks good.

Good spot by Jorn that ImageFileCreator wasn't closing.

**Pull request closed without merging**